### PR TITLE
Add API endpoint to check consensus status(Reopened) [ECR-1649]

### PR DESCRIPTION
**Pull request by SergeiMal**
Check for active validators count

### CODE REVIEW

File: exonum/src/blockchain/service.rs [hunk 1]
```diff
@@ -323,6 +323,9 @@ pub struct ApiNodeState {
     // TODO: Update on event? (ECR-1632)
     peers_info: HashMap<SocketAddr, PublicKey>,
     is_enabled: bool,
+    is_validator: bool,
+    majority_count:usize,
```
Comment (vitvakatu) [inline exonum/src/blockchain/service.rs]: missing space

File: exonum/src/blockchain/service.rs [hunk 2]
```diff
@@ -395,19 +398,42 @@ impl SharedNodeState {
...
+            lock.peers_info.insert(c.addr(), *p);
         }
+        for v in state.validators().iter() {
+            lock.validators.push(*v);
+        }
+    }
+
+    /// Returns a boolean value which indicates whether the consensus is
+    /// Returns the majority count from the current "State"
+    /// of a blockchain node.
```
Comment (vitvakatu) [inline exonum/src/blockchain/service.rs]: Docs are messed up

File: exonum/src/blockchain/service.rs [hunk 3]
```diff
@@ -395,19 +398,42 @@ impl SharedNodeState {
...
+    pub fn is_consensus(&self) -> bool {
+        let lock = self.state.read()
+            .expect("Expected read lock.");
+
+        let mut active_validators :usize = 0;
```
Comment (vitvakatu) [inline exonum/src/blockchain/service.rs]: wrong spacing

File: exonum/src/blockchain/service.rs [hunk 4]
```diff
@@ -395,19 +398,42 @@ impl SharedNodeState {
...
+        for (_s, p) in lock.peers_info.iter() {
+            if lock.validators.iter()
+                .any(|x| x.consensus_key == *p) {
+                active_validators = active_validators.saturating_add(1);
```
Comment (vitvakatu) [inline exonum/src/blockchain/service.rs]: I like safety, but `active_validators` is `usize` and I don't think it's actually realistic to use `saturating_add`.

File: exonum/src/blockchain/service.rs [hunk 5]
```diff
@@ -395,19 +398,42 @@ impl SharedNodeState {
...
+            }
+        }
+
+        if lock.is_validator {
+            // Peers list doesn't include current node address, so we have to increment its length.
+            // E.g. if we have 3 items in peers list, it means that we have 4 nodes overall.
+            active_validators = active_validators.saturating_add(1);
+        }
+
+        lock.majority_count > 0 && active_validators >= lock.majority_count
```
Comment (vitvakatu) [inline exonum/src/blockchain/service.rs]: Why do we need to check for `majority_count > 0`?

File: testkit/tests/system_api.rs [hunk 6]
```diff
@@ -52,3 +53,12 @@ fn test_network() {
     assert_eq!(info.protocol_version, PROTOCOL_MAJOR_VERSION);
     assert!(info.services.is_empty());
 }
+
+#[test]
+fn test_consensus_status() {
```
Comment (vitvakatu) [inline testkit/tests/system_api.rs]: I suppose, one validator isn't the right test case here. And you need to check for `false` values too.

File: exonum/src/blockchain/service.rs [hunk 7]
```diff
@@ -395,19 +398,42 @@ impl SharedNodeState {
...
+    /// Returns the majority count from the current "State"
+    /// of a blockchain node.
+    pub fn is_consensus(&self) -> bool {
+        let lock = self.state.read()
+            .expect("Expected read lock.");
+
+        let mut active_validators :usize = 0;
+        for (_s, p) in lock.peers_info.iter() {
+            if lock.validators.iter()
+                .any(|x| x.consensus_key == *p) {
```
Comment (popzxc) [inline exonum/src/blockchain/service.rs]: Personally I find expression `x.consensus_key == *p` very confusing. Is it possible to make it more readable?

File: exonum/src/blockchain/service.rs [hunk 8]
```diff
@@ -395,19 +398,42 @@ impl SharedNodeState {
...
         }
+        for v in state.validators().iter() {
+            lock.validators.push(*v);
+        }
+    }
+
+    /// Returns a boolean value which indicates whether the consensus is
+    /// Returns the majority count from the current "State"
+    /// of a blockchain node.
+    pub fn is_consensus(&self) -> bool {
```
Comment (popzxc) [inline exonum/src/blockchain/service.rs]: Is it a good name?

**Comment by popzxc** [comment]
General question: why is this PR was made from the already merged branch? 
Shouldn't it contain only one commit?

File: exonum/src/blockchain/service.rs (hunk 7 above)
Comment (vitvakatu) [reply exonum/src/blockchain/service.rs]: I think the only way would be to rename `p` to something like `peer_key`

Comment (popzxc) [reply exonum/src/blockchain/service.rs]: And `x` to `peer`, I guess.

File: exonum/src/blockchain/service.rs (hunk 5 above)
Comment (SergeiMal) [reply exonum/src/blockchain/service.rs]: At the beginning, just before the state uptaded, active_validators = 0 and majority_count = 0, and as result is_consensus will give true.

File: testkit/tests/system_api.rs (hunk 6 above)
Comment (SergeiMal) [reply testkit/tests/system_api.rs]: It will be changed during ECR-1665

File: exonum/src/blockchain/service.rs (hunk 5 above)
Comment (popzxc) [reply exonum/src/blockchain/service.rs]: I guess we should at least add a comment about it.

File: exonum/src/blockchain/service.rs (hunk 7 above)
Comment (vitvakatu) [reply exonum/src/blockchain/service.rs]: `x` should be `validator`, I think

Comment (popzxc) [reply exonum/src/blockchain/service.rs]: Sure.

**Comment by SergeiMal** [comment]
The PR closed. All changes (+ fixes according the comments ) will be committed with new separated PR

**Pull request closed without merging**